internal/shellgen/flake_package.go

```diff
@@ -0,0 +1,66 @@
+package shellgen
+
+import (
+	"path/filepath"
+	"strings"
+
+	"go.jetpack.io/devbox/internal/boxcli/featureflag"
+	"go.jetpack.io/devbox/internal/nix"
+)
+
+type flakePackage struct {
+	Name      string
+	FromStore string
+	FromPath  string
+}
+
+func newFlakePackage(pkg *nix.Package) (*flakePackage, error) {
+	// nixosCacheURL is where we fetch package binaries from
+	const nixosCacheURL = "https://cache.nixos.org"
+
+	// flakePackages only support versioned packages that have a system info
+	sysInfo := pkg.SystemInfo()
+	if sysInfo == nil {
+		return nil, nil
+	}
+
+	attributePath, err := pkg.PackageAttributePath()
+	if err != nil {
+		return nil, err
+	}
+
+	storeDir := strings.Join([]string{sysInfo.FromHash, sysInfo.StoreName, sysInfo.StoreVersion}, "-")
+	return &flakePackage{
+		Name:      attributePath,
+		FromStore: nixosCacheURL,
+		FromPath:  filepath.Join("/nix/store", storeDir),
+		// TODO add ToPath:
+	}, nil
+}
+
+func flakePackages(devbox devboxer) ([]*flakePackage, error) {
```

Comment:
I know this wasn't added in your PR, but the number of interfaces we have is making the code really hard to follow.